language/move-prover/bytecode-to-boogie/src/bytecode_instrs.bpl

```diff
@@ -15,16 +15,67 @@ function {:constructor} Path(p: [int]Edge, size: int): Path;
 
 const DefaultPath: [int]Edge;
 
+type {:datatype} TypeValue;
+function {:constructor} BooleanType() : TypeValue;
+function {:constructor} IntegerType() : TypeValue;
+function {:constructor} AddressType() : TypeValue;
+function {:constructor} ByteArrayType() : TypeValue;
+function {:constructor} StrType() : TypeValue;
+function {:constructor} ReferenceType(t: TypeValue) : TypeValue;
+function {:constructor} VectorType(t: TypeValue) : TypeValue;
+// StructTypes caries the arity (# of fields) so we can enumerate fields (which are ints)
+function {:constructor} StructType(name: TypeName, arity: int, ts: [FieldName]TypeValue) : TypeValue;
+
 type {:datatype} Value;
 function {:constructor} Boolean(b: bool): Value;
 function {:constructor} Integer(i: int): Value;
 function {:constructor} Address(a: Address): Value;
 function {:constructor} ByteArray(b: ByteArray): Value;
 function {:constructor} Str(a: String): Value;
-function {:constructor} Map(m: [Edge]Value): Value;
-function {:constructor} Vector(v: [Edge]Value, l: int): Value;
+// Currently empty vectors of arbitrary type will be identified as equal. If this causes a problem,
+// we need to add the element's TypeValue to the vector.
+function {:constructor} Vector(v: [int]Value, l: int): Value;
+// Currently structural equal structs are identified. We need to add the type name below if this causes a problem.
+function {:constructor} Struct(v: [FieldName]Value): Value;
+
+// Axiomatic definition of type test and equality.
+function has_type(t: TypeValue, v: Value): bool;
```

Comment:
Just curious: Is has_type used anywhere in the translation?

Reply:
Yep, about nearly everywhere. The current generator extensively generates `assume has_type(T, v)` (previously it did `assume is#T(v)` which doesn't work with generics. The rationale seems to be to let boogie know and benefit from that the code is already type checked. I was thinking of an alternative, strongly typed encoding but we can keep that for later.